File: Matrix.java
class Matrix {
    public boolean searchMatrix(int[][] mat, int k) {
        int m =mat.length;
        int n=mat[0].length;
        int l = 0, r = m - 1;
        while (l <= r)
        {
            int mid = l + (r - l) / 2;

            if (k >= mat[mid][0] && k <= mat[mid][n - 1])
                return searchRow(mat[mid], n, k);

            if (k < mat[mid][0])
                r = mid - 1;
            else
                l = mid + 1;
        }
        return false;
    }
    boolean searchRow(int M[], int n, int k)
    {
        int l = 0, r = n - 1;
        while (l <= r)
        {
            int mid = l + (r - l) / 2;

            if (M[mid] == k)
                return true;

            if (k < M[mid])
                r = mid - 1;
            else
                l = mid + 1;
        }
        return false;
    }

}

File: Matrix.py
# Time Complexity: O(n)
# Space Complexity: O(1)
class Solution(object):
    def searchMatrix(self, matrix, target):

        # Getting the row and column size
        row = len(matrix)
        column = 0
        if len(matrix) != 0:
            column = len(matrix[0])

        row_counter = 0

        # Searching linearly in the matrix for the given target
        while column >= 0 and row_counter < row:
            if matrix[row_counter][column - 1] == target:
                return True
            if matrix[row_counter][column - 1] > target:
                column -= 1
            else:
                row_counter += 1

        return False

        """
        :type matrix: List[List[int]]
        :type target: int
        :rtype: bool
        """
